File: pams_python/silhouette.py
import numpy as np
from pam_builder import PamBuilder


def calc_silhouette(dist_matrix, clusters):
    num_points = dist_matrix.shape[0]
    a_vals = np.zeros(num_points)
    b_vals = np.full(num_points, np.inf)

    point_indices = np.arange(num_points)

    for i, cluster in enumerate(clusters):
        in_cluster = np.isin(point_indices, cluster)
        cluster_dist = dist_matrix[np.ix_(in_cluster, in_cluster)]
        a_vals[in_cluster] = np.mean(np.ma.masked_where(cluster_dist == np.inf, cluster_dist), axis=1)

        # Calculate b_i values by comparing to other clusters
        for j, other_cluster in enumerate(clusters):
            if i != j:
                # Boolean mask for the points in the other cluster
                other_in_cluster = np.isin(point_indices, other_cluster)
                # Calculate mean distance from points in current cluster to points in other cluster
                inter_cluster_dist = np.mean(dist_matrix[np.ix_(in_cluster, other_in_cluster)], axis=1)
                # Update b_i values only if the calculated mean is lower than the current b_i
                b_vals[in_cluster] = np.minimum(b_vals[in_cluster], inter_cluster_dist)

    # Calculate the silhouette values
    silhouette_vals = (b_vals - a_vals) / np.maximum(a_vals, b_vals)

    # Calculate the overall mean silhouette score
    silhouette_score = np.mean(silhouette_vals)

    return silhouette_score


def create_pam_clusters(dist_matrix):
    DECREASE_THRESHOLD = 5

    best_clusters = []
    best_silhouette_value = -1
    prev_silhouette_value = -np.inf
    decrease_count = 0

    pam_builder = PamBuilder(dist_matrix)

    for k in range(2, dist_matrix.shape[0] + 1):
        clusters, medoids = pam_builder.pam(k)
        silhouette_value = calc_silhouette(dist_matrix, clusters)
        print(f"Silhouette for k={k}: {silhouette_value}")

        if silhouette_value > best_silhouette_value:
            best_silhouette_value = silhouette_value
            best_clusters = clusters

        if silhouette_value < prev_silhouette_value:
            decrease_count += 1
        else:
            decrease_count = 0

        if decrease_count == DECREASE_THRESHOLD:
            print(f"Stopping at k={k} due to {DECREASE_THRESHOLD} consecutive decreases in silhouette value.")
            break

        prev_silhouette_value = silhouette_value

    return best_clusters

File: pams/Silhouette.cpp

#include <cmath>
#include <execution>
#include <iostream>
#include <limits>
#include <numeric>

#include "Pam.h"
#include "Silhouette.h"

std::vector<std::vector<size_t>> createPamClusters(const Matrix& distMatrix){
  const int DECREASE_THRESHOLD = 5;
 

  std::vector<PamResult> pamResults();


  std::vector<std::vector<size_t>> resultClusters;
  double bestSilhouetteValue = -1.0;
  double prevSilhouetteValue = std::numeric_limits<double>::lowest();
  int decreaseCount = 0;
  int localMaxK = 1;

  PamBuilder pamBuilder(distMatrix);

  for (int k = 2;; ++k) {
    auto [clusters, medoids] = pamBuilder.pam(k);
    double curSilhouetteValue = calcSilhouette(distMatrix, clusters, medoids);
    std::cout << "Silhouette for k=" << k << ": " << curSilhouetteValue << "\n";

    if (curSilhouetteValue > bestSilhouetteValue) {
      bestSilhouetteValue = curSilhouetteValue;
      resultClusters = clusters;
      localMaxK = k;
    }

    if (curSilhouetteValue < prevSilhouetteValue) {
      decreaseCount++;
    } else {
      decreaseCount = 0;
    }

    if (decreaseCount == DECREASE_THRESHOLD) {
      std::cout << "Stopping at k=" << k << " due to " << DECREASE_THRESHOLD
                << " consecutive decreases in silhouette value.\n";
      break;
    }

    prevSilhouetteValue = curSilhouetteValue;
  }

  std::cout << "Best k found: " << localMaxK
            << " with silhouette: " << bestSilhouetteValue << "\n";
  return resultClusters;
}


// Impelemtation acording to https://en.wikipedia.org/wiki/Silhouette_(clustering)
double calcSilhouette(const Matrix& distMatrix,
                      const std::vector<std::vector<size_t>>& clusters,
                      const std::vector<size_t>& medoids) {
  size_t numPoints = distMatrix.size();
  std::vector<double> silhouetteVals(numPoints, 0.0);

  for (size_t i = 0; i < clusters.size(); ++i) {
    for (size_t j = 0; j < clusters[i].size(); ++j) {
      size_t point = clusters[i][j];

      // Calculate a(i)
      double a = 0.0;
      if (clusters[i].size() > 1) {
        for (size_t k = 0; k < clusters[i].size(); ++k) {
          if (point != clusters[i][k]) {
            a += distMatrix[point][clusters[i][k]];
          }
        }
        a /= (clusters[i].size() - 1);
      }

      // Calculate b(i)
      double b = std::numeric_limits<double>::max();
      for (size_t m = 0; m < clusters.size(); ++m) {
        if (m != i) {
          double avgDist = 0.0;
          for (size_t k = 0; k < clusters[m].size(); ++k) {
            avgDist += distMatrix[point][clusters[m][k]];
          }
          avgDist /= clusters[m].size();
          if (avgDist < b) {
            b = avgDist;
          }
        }
      }

      // Calculate s(i)
      double s = (b - a) / std::max(a, b);
      silhouetteVals[point] = s;
    }
  }

  // Calculate the mean silhouette score
  double silhouetteScore =
      std::reduce(silhouetteVals.begin(), silhouetteVals.end()) / numPoints;

  return silhouetteScore;
}
